File: tests/solutions_js/find_median_from_data_stream.js
class MedianFinder {
  constructor() {
    this.lower = []; // max-heap (stored as negated values)
    this.upper = []; // min-heap
  }

  addNum(num) {
    this._heapPush(this.lower, -num);

    // Ensure max of lower <= min of upper
    if (this.lower.length && this.upper.length && -this.lower[0] > this.upper[0]) {
      const val = -this._heapPop(this.lower);
      this._heapPush(this.upper, val);
    }

    // Rebalance sizes
    if (this.lower.length > this.upper.length + 1) {
      const val = -this._heapPop(this.lower);
      this._heapPush(this.upper, val);
    } else if (this.upper.length > this.lower.length) {
      const val = this._heapPop(this.upper);
      this._heapPush(this.lower, -val);
    }
  }

  findMedian() {
    if (this.lower.length > this.upper.length) {
      return -this.lower[0];
    }
    return (-this.lower[0] + this.upper[0]) / 2;
  }

  _heapPush(heap, val) {
    heap.push(val);
    let i = heap.length - 1;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (heap[parent] <= heap[i]) break;
      [heap[parent], heap[i]] = [heap[i], heap[parent]];
      i = parent;
    }
  }

  _heapPop(heap) {
    const top = heap[0];
    const last = heap.pop();
    if (heap.length > 0) {
      heap[0] = last;
      let i = 0;
      while (true) {
        let smallest = i;
        const left = 2 * i + 1, right = 2 * i + 2;
        if (left < heap.length && heap[left] < heap[smallest]) smallest = left;
        if (right < heap.length && heap[right] < heap[smallest]) smallest = right;
        if (smallest === i) break;
        [heap[i], heap[smallest]] = [heap[smallest], heap[i]];
        i = smallest;
      }
    }
    return top;
  }
}

File: tests/solutions_cpp/find_median_from_data_stream.cpp
#include "json_helper.h"

class MedianFinder {
public:
    priority_queue<int> lower;                                   // max-heap (lower half)
    priority_queue<int, vector<int>, greater<int>> upper;        // min-heap (upper half)

    MedianFinder() {}

    void addNum(int num) {
        lower.push(num);
        // Ensure ordering: every element in lower <= every element in upper
        if (!lower.empty() && !upper.empty() && lower.top() > upper.top()) {
            int v = lower.top(); lower.pop();
            upper.push(v);
        }
        // Balance sizes: lower may have at most 1 more than upper
        if ((int)lower.size() > (int)upper.size() + 1) {
            int v = lower.top(); lower.pop();
            upper.push(v);
        } else if ((int)upper.size() > (int)lower.size()) {
            int v = upper.top(); upper.pop();
            lower.push(v);
        }
    }

    double findMedian() {
        if (lower.size() > upper.size()) return (double)lower.top();
        return ((double)lower.top() + (double)upper.top()) / 2.0;
    }
};

int main() {
    auto args = parse_args();
    auto& outer = args[0].to_obj();
    auto& ops = outer.at("ops").to_arr();
    auto& opArgs = outer.at("args").to_arr();

    MedianFinder mf;
    MultiOpResult results;
    results.add_null(); // constructor
    for (size_t i = 1; i < ops.size(); i++) {
        const string& op = ops[i].to_str();
        auto& a = opArgs[i].to_arr();
        if (op == "addNum") {
            mf.addNum(a[0].to_int());
            results.add_null();
        } else if (op == "findMedian") {
            results.add_double(mf.findMedian());
        }
    }
    cout << results.to_json_str() << endl;
    return 0;
}
